trust-quorum/tests/coordinator.rs

```diff
@@ -0,0 +1,646 @@
+                // We expect retries for all members that the coordinator
+                // has not received acks for.
+                let expected: BTreeSet<_> =
+                    members.difference(acked_members).collect();
+                for envelope in &outbox {
+                    prop_assert!(expected.contains(&envelope.to));
+                }
+            } else {
+                // We aren't waiting on acks, so won't retry sending prepares
+                prop_assert!(outbox.is_empty());
+            }
+        }
+
+        // Put any output messages onto the network
+        self.send(outbox.into_iter());
+
+        Ok(())
+    }
+
+    fn reply_to_prepare_msg(
+        &mut self,
+        from: PlatformId,
+        msg: PrepareMsg,
+    ) -> Result<(), TestCaseError> {
+        // We always reply to a `Prepare` with a `PrepareAck`.
+        //
+        // We are testing the coordinator, and assume it is sending
+        // `Prepares` that are up to date, and expecting back appropriate acks.
+        //
+        // In any case, we don't keep enough state at the fake follower replicas
+        // to check this.
+        let reply = PeerMsg::PrepareAck(msg.config.epoch);
+        let mut outbox = Vec::new();
+        let output = self.sut.node.handle(
+            self.model.now,
+            &mut outbox,
+            from.clone(),
+            reply,
+        );
+        prop_assert!(output.is_none());
+        prop_assert!(outbox.is_empty());
+
+        // Also update the model state
+        self.model.ack_prepare(from.clone(), msg.config.epoch);
+
+        // Ensure that if the SUT is waiting for prepares, that the ack
+        // is accounted for.
+        if let Some(acks) = self.model.acked_prepares() {
+            prop_assert!(acks.contains(&from));
+        }
+
+        Ok(())
+    }
+
+    /// Ensure that the output of `Node::coordinate_reconfiguration`
+    /// is valid given the `TestState`.
+    ///
+    /// This is essentially a "postcondition" check.
+    fn assert_persistent_state_after_coordinate_reconfiguration(
+        &self,
+        persistent_state: &PersistentState,
+    ) -> Result<(), TestCaseError> {
+        let sut = &self.sut;
+        let msg = &self
+            .model
+            .coordinator_state
+            .as_ref()
+            .ok_or(TestCaseError::fail(
+                "Model should have a coordinator state",
+            ))?
+            .msg;
+
+        prop_assert!(persistent_state.lrtq.is_none());
+        prop_assert_eq!(
+            &sut.persistent_state.commits,
+            &persistent_state.commits
+        );
+        prop_assert!(persistent_state.decommissioned.is_none());
+        prop_assert_eq!(
+            sut.persistent_state.prepares.len() + 1,
+            persistent_state.prepares.len()
+        );
+
+        prop_assert_eq!(
+            persistent_state.last_prepared_epoch().unwrap(),
+            msg.epoch
+        );
```

Comment:
A general thing I'd mention here is that often when applying a series of operations, it can be helpful to track things like which index the operation fails on. In the past I've used anyhow errors at this level, attaching context about the errors at higher levels.

Reply:
That's a good idea. I'll keep it in mind.